### PR TITLE
Hide the value of sensitive query parameters in log

### PR DESCRIPTION
**Pull request by MinhDang685**
For logging interceptor: add an option to hide the value of sensitive query parameters in the request and response URL. 
Inspired by the implementation of header redaction in `okhttp3.logging.HttpLoggingInterceptor`

**To use**
Call the method `redactQueryParams(the list of query parameters to hide value, separated by commas)` from a `okhttp3.logging.HttpLoggingInterceptor` instance. For example, to hide the value of "user" and "password" query parameters, use `interceptor.redactQueryParams("user", "password")`.
- Original URL: `http://localhost:8000/api/login?user=user1&authentication=basic&password=confidential_password"`
- With the method applied: `http://localhost:8000/api/login?user=██&authentication=basic&password=██"`

### CODE REVIEW

**Comment by yschimke** [comment]
This is against the resolution of https://github.com/square/okhttp/issues/7476

> But I continue to believe that HttpLoggingInterceptor is not designed for production use. Speculative logging is too noisy. For development I love Chucker; for production I’d like something that does spans like Datadog or Chronosphere.

As it's a functional PR, I'm inclined to accept it.

Will wait for a final take by @swankjesse

**Review by yschimke — APPROVED** [review]

File: okhttp-logging-interceptor/src/test/java/okhttp3/logging/HttpLoggingInterceptorTest.kt [hunk 1]
```diff
@@ -903,6 +903,51 @@ class HttpLoggingInterceptorTest {
       .assertNoMoreLogs()
   }
 
+  @Test
+  fun sensitiveQueryParamsAreRedacted() {
```
Comment (yschimke) [inline okhttp-logging-interceptor/src/test/java/okhttp3/logging/HttpLoggingInterceptorTest.kt:907]: Can you add a test for duplicate query params, and also retaining the order.

File: okhttp-logging-interceptor/src/main/kotlin/okhttp3/logging/HttpLoggingInterceptor.kt [hunk 2]
```diff
@@ -312,6 +322,32 @@ class HttpLoggingInterceptor
       return response
     }
 
+    private fun sanitizeUrl(url: HttpUrl): String {
```
Comment (yschimke) [inline okhttp-logging-interceptor/src/main/kotlin/okhttp3/logging/HttpLoggingInterceptor.kt:325]: Actually, I think this needs rework.

queryParameterNames is a unique Set.

Maybe something like

```
    private fun sanitizeUrl(url: HttpUrl): String {
      if (queryParamsNameToRedact.isEmpty() || url.querySize == 0) {
        return url.toString()
      }

      val newUrl = url.newBuilder().query(null).apply {
        for (i in 0 until url.querySize) {
          val parameterName = url.queryParameterName(i)
          val newValue = if (parameterName in queryParamsNameToRedact) "██" else url.queryParameterValue(i)

          addEncodedQueryParameter(parameterName, newValue)
        }
      }.toString()
    }
```

Comment (MinhDang685) [reply okhttp-logging-interceptor/src/main/kotlin/okhttp3/logging/HttpLoggingInterceptor.kt:325]: Thanks for your suggestion; this one is definitely better than my string concatenation solution.

File: okhttp-logging-interceptor/src/main/kotlin/okhttp3/logging/HttpLoggingInterceptor.kt [hunk 3]
```diff
@@ -312,6 +322,20 @@ class HttpLoggingInterceptor
       return response
     }
 
+    internal fun sanitizeUrl(url: HttpUrl): String {
```
Comment (swankjesse) [inline okhttp-logging-interceptor/src/main/kotlin/okhttp3/logging/HttpLoggingInterceptor.kt:325]: Rename to redactUrl please

File: okhttp-logging-interceptor/src/test/java/okhttp3/logging/HttpLoggingInterceptorTest.kt [hunk 4]
```diff
@@ -903,6 +903,99 @@ class HttpLoggingInterceptorTest {
...
+        .newCall(
+          request()
+            .build(),
+        )
+        .execute()
+    response.body.close()
+    val sanitizedUrl = networkInterceptor.sanitizeUrl(url)
+    val sanitizedUrlPattern = sanitizedUrl.replace("?", """\?""")
+    applicationLogs
+      .assertLogEqual("--> GET $sanitizedUrl")
```
Comment (swankjesse) [inline okhttp-logging-interceptor/src/test/java/okhttp3/logging/HttpLoggingInterceptorTest.kt:990]: Please use the word `redactedUrl` here.

File: okhttp-logging-interceptor/src/test/java/okhttp3/logging/HttpLoggingInterceptorTest.kt [hunk 5]
```diff
@@ -903,6 +903,99 @@ class HttpLoggingInterceptorTest {
...
+      .assertLogMatch(Regex("""<-- 200 OK $sanitizedUrlPattern \(\d+ms, \d+-byte body\)"""))
+      .assertNoMoreLogs()
+    networkLogs
+      .assertLogEqual("--> GET $sanitizedUrl http/1.1")
+      .assertLogMatch(Regex("""<-- 200 OK $sanitizedUrlPattern \(\d+ms, \d+-byte body\)"""))
+      .assertNoMoreLogs()
+  }
+
+  @Test
+  fun preserveQueryParamsAfterSanitized() {
```
Comment (swankjesse) [inline okhttp-logging-interceptor/src/test/java/okhttp3/logging/HttpLoggingInterceptorTest.kt:951]: Please use the word `Redacted` instead of `Sanitzed` throughout!

**Review by swankjesse — APPROVED** [review]
Works for me. I’ve got some naming changes I’d like you to make before I merge it!

File: okhttp-logging-interceptor/src/test/java/okhttp3/logging/HttpLoggingInterceptorTest.kt [hunk 6]
```diff
@@ -903,6 +903,101 @@ class HttpLoggingInterceptorTest {
       .assertNoMoreLogs()
   }
 
+  @Suppress("INVISIBLE_MEMBER", "INVISIBLE_REFERENCE")
```
Comment (yschimke) [inline okhttp-logging-interceptor/src/test/java/okhttp3/logging/HttpLoggingInterceptorTest.kt:906]: @MinhDang685 we run these tests in android also, which means they need to workaround internal methods. So I added this.
